crates/trie/common/src/proof_fetching.rs

```diff
@@ -0,0 +1,351 @@
+//! Merkle trie proofs.
+
+use crate::{ChunkedMultiProofTargets, MultiProofTargets};
+use alloy_primitives::{
+    map::{B256Map, B256Set},
+    B256,
+};
+use core::cmp::Ordering;
+
+/// Represents a single proof target. This is a helper struct to be used when flattening a
+/// [`TargetsToFetch`], similar to an `Option<B256>` but preserving information about whether or
+/// not the account proof should be fetched.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum SingleProofTarget {
+    /// Account proof target, no storage proofs.
+    Account,
+    /// Storage proof target.
+    StorageOnly(B256),
+    /// Account with storage proof target.
+    AccountWithStorage(B256),
+}
+
+impl SingleProofTarget {
+    /// Returns an [`Option`] containing the storage slot if a storage slot exists for the target.
+    pub fn storage_slot(&self) -> Option<B256> {
+        match self {
+            Self::StorageOnly(slot) | Self::AccountWithStorage(slot) => Some(*slot),
+            Self::Account => None,
+        }
+    }
+}
+
+impl PartialOrd for SingleProofTarget {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for SingleProofTarget {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let self_slot = self.storage_slot();
+        let other_slot = other.storage_slot();
+        self_slot.cmp(&other_slot)
+    }
+}
+
+/// Represents proof targets that should be fetched.
+///
+/// This is meant to be constructed by comparing two [`MultiProofTargets`] instances, where one is
+/// the already-fetched proof targets, and the other is the new set of proof targets.
+///
+/// For example:
+/// ```
+/// use alloy_primitives::{map::B256Map, B256};
+/// use reth_trie_common::{MultiProofTargets, TargetsToFetch};
+///
+/// let fetched = MultiProofTargets::accounts([
+///     B256::from([1; 32]),
+///     B256::from([2; 32]),
+///     B256::from([3; 32]),
+/// ]);
+/// let new = MultiProofTargets::accounts([
+///     B256::from([1; 32]),
+///     B256::from([2; 32]),
+///     B256::from([4; 32]),
+/// ]);
+///
+/// let targets_to_fetch = TargetsToFetch::from_difference(&fetched, new);
+/// let expected =
+///     TargetsToFetch::from_account_targets([(B256::from([4; 32]), Default::default())]);
+///
+/// assert_eq!(targets_to_fetch, expected);
+/// ```
+#[derive(Clone, PartialEq, Eq, Debug, Default)]
+pub struct TargetsToFetch {
```

Comment:
I think introducing another type for the sake of separating account and storage targets is not needed, I would rather change the `MultiProofTargets` definition, and basically have the same interface as before but with the new input type for proofs.

@rkrasiuk wdyt?

Reply:
i agree, current impl seems to verbose